File: app/src/platform/runtimes.ts
import { useCallback, useEffect, useState } from "react";
import { request } from "./daemon";
import type { RunningPane } from "./generated/RunningPane";
import type { TerminalView } from "./generated/TerminalView";

/**
 * The terminal Dex opens: `windows` (PowerShell) or `wsl:<distro>`. One
 * choice for the whole app (`pane.terminal`): every new pane and every agent
 * spawned opens there, and when Dex starts, every pane does. A pane already
 * running keeps its shell unless the owner has it restarted (`restartChoices`).
 */

/** The distro of a WSL terminal; null for Windows or none. */
export function distroOf(runtime: string | undefined): string | null {
  return runtime?.startsWith("wsl:") ? runtime.slice(4) : null;
}

/** A terminal as the owner reads it: "PowerShell", "Ubuntu (WSL)". */
export function runtimeLabel(runtime: string): string {
  if (runtime === "windows") return "PowerShell";
  const distro = distroOf(runtime);
  return distro ? `${distro} (WSL)` : runtime;
}

/**
 * Whether this window can close a pane's shell and start it again in `next`:
 * it is running here - not exited, not shown in another window - and in
 * another terminal. Whether it should is the owner's call.
 */
export function switchNow(
  shell: { spawned: boolean; dead: boolean; away: boolean; runtime?: string },
  next: string,
): boolean {
  return shell.spawned && !shell.dead && !shell.away && shell.runtime !== undefined && shell.runtime !== next;
}

/** One pane in the "restart these in the new terminal?" list. */
export interface RestartChoice {
  pane: string;
  title: string;
  detail: string;
  busy: boolean;
  /** Ticked to begin with: a plain shell is; one that may be busy is not. */
  ticked: boolean;
}

/**
 * The panes to offer to restart after a terminal is chosen: those running
 * elsewhere that this window can restart (not ones shown in another window).
 * Plain shells start ticked; a pane where something may be running does not,
 * and says so - restarting it would end that. A pane with a live agent is
 * busy whatever the process table says, and names the agent.
 */
export function restartChoices(
  running: readonly RunningPane[],
  canRestart: (pane: string) => boolean,
  agentIn: (pane: string) => string | null = () => null,
): RestartChoice[] {
  return running
    .filter((pane) => canRestart(pane.pane))
    .map((pane) => {
      const agent = agentIn(pane.pane);
      const busy = pane.busy || agent !== null;
      const why = agent !== null ? ` · agent ${agent} runs here` : busy ? " · something may be running in it" : "";
      return {
        pane: pane.pane,
        title: pane.label ?? pane.cwd.split(/[\\/]/).filter(Boolean).at(-1) ?? pane.cwd,
        detail: `${pane.workspace} · ${runtimeLabel(pane.runtime)}${why}`,
        busy,
        ticked: !busy,
      };
    });
}

/**
 * Of the panes the owner ticked, the ones to restart now, asked again at the
 * moment of the click: one no longer running elsewhere is left, and one that
 * was plain when offered but is busy now is skipped - something started in it
 * since. A pane offered as busy and ticked anyway is the owner's decision.
 */
export function stillToRestart(
  ticked: ReadonlySet<string>,
  offered: readonly RestartChoice[],
  now: readonly RestartChoice[],
): { restart: string[]; skipped: string[] } {
  const current = new Map(now.map((choice) => [choice.pane, choice]));
  const restart: string[] = [];
  const skipped: string[] = [];
  for (const choice of offered) {
    if (!ticked.has(choice.pane)) continue;
    const fresh = current.get(choice.pane);
    if (!fresh) continue;
    if (fresh.busy && !choice.busy) skipped.push(choice.title);
    else restart.push(choice.pane);
  }
  return { restart, skipped };
}

/**
 * What the terminal picker offers: every terminal there is, or nothing where
 * Windows is all there is. A chosen distro that has since been uninstalled
 * stays in the list, marked, so the owner can see it and choose away from it.
 */
export function terminalChoices(view: TerminalView): { value: string; label: string }[] {
  const gone = !view.runtimes.includes(view.terminal);
  if (view.runtimes.length < 2 && !gone) return [];
  const choices = view.runtimes.map((value) => ({ value, label: runtimeLabel(value) }));
  if (gone) choices.push({ value: view.terminal, label: `${runtimeLabel(view.terminal)} - not installed` });
  return choices;
}

/**
 * The chosen terminal and the choices, and a way to choose; null until asked.
 * Asked again whenever `generation` changes - the setup panel's checks - so a
 * distro installed meanwhile, or a first ask that failed, is caught up with.
 */
export function useTerminal(generation?: unknown): {
  view: TerminalView | null;
  choose: (runtime: string) => Promise<TerminalView>;
} {
  const [view, setView] = useState<TerminalView | null>(null);
  useEffect(() => {
    let live = true;
    void request<TerminalView>("pane.terminal", {}).then(
      (answer) => live && setView(answer),
      () => {},
    );
    return () => {
      live = false;
    };
  }, [generation]);
  const choose = useCallback(async (runtime: string) => {
    const answer = await request<TerminalView>("pane.terminal", { terminal: runtime });
    setView(answer);
    return answer;
  }, []);
  return { view, choose };
}

File: crates/dex-core/src/features/workspace/runtimes.rs
//! Which terminal Dex opens: PowerShell (or the configured shell) on Windows,
//! or a WSL distro's shell (`platform/wsl.rs`). One choice for the whole app,
//! kept in `app_state`: every new pane opens there - a new workspace's, a
//! split, a spawned agent's - and when Dex starts, every pane does. A pane
//! records the terminal its shell starts in as its `runtime`, so one already
//! running keeps going where it is when the choice changes.

use std::collections::HashMap;
use std::path::Path;

use dex_protocol::pane::{TerminalArgs, TerminalView};

use super::model::WorkspaceError;
use super::store;
use crate::app::AppState;
use crate::platform::proctree;
use crate::platform::wsl::{self, Runtime};

/// Checks a terminal against the distros installed, and names the distro as
/// `wsl.exe` does (it matches names regardless of case). Pure: `installed` is
/// what `wsl.exe` listed.
pub(super) fn check(text: &str, installed: &[String]) -> Result<String, WorkspaceError> {
    match Runtime::parse(text).map_err(WorkspaceError::InvalidRuntime)? {
        Runtime::Windows => Ok(Runtime::Windows.to_string()),
        Runtime::Wsl(distro) => match installed
            .iter()
            .find(|name| name.eq_ignore_ascii_case(&distro))
        {
            Some(name) => Ok(Runtime::Wsl(name.clone()).to_string()),
            None => Err(WorkspaceError::NoSuchDistro {
                distro,
                installed: installed.to_vec(),
            }),
        },
    }
}

/// Whether a pane in this folder stays on Windows when Dex's terminal is WSL:
/// a worktree Windows git made records where its repository is as a `C:/...`
/// path, which Linux git cannot follow, so git there would say "not a git
/// repository". Pure: `git_file` is the folder's `.git`, when that is a file.
pub(super) fn windows_only(git_file: Option<&str>) -> bool {
    git_file
        .and_then(|text| text.trim().strip_prefix("gitdir:"))
        .map(str::trim)
        .is_some_and(|dir| {
            let bytes = dir.as_bytes();
            bytes.len() >= 2 && bytes[0].is_ascii_alphabetic() && bytes[1] == b':'
        })
}

fn git_file(cwd: &str) -> Option<String> {
    let path = Path::new(cwd).join(".git");
    path.is_file()
        .then(|| std::fs::read_to_string(path).ok())
        .flatten()
}

/// Moves terminal panes to `runtime`: every one, or with `idle_only` those
/// whose shell is not running. Returns how many moved.
pub(super) async fn move_panes(
    state: &AppState,
    runtime: String,
    idle_only: bool,
) -> Result<usize, WorkspaceError> {
    let panes = state.db.call(|conn| store::terminal_panes(conn)).await?;
    let to_wsl = matches!(Runtime::parse(&runtime), Ok(Runtime::Wsl(_)));
    let ids: Vec<String> = panes
        .into_iter()
        .filter(|(id, _, _)| !idle_only || state.pty.shell_pid(id).is_none())
        .filter(|(_, cwd, _)| !(to_wsl && windows_only(git_file(cwd).as_deref())))
        .map(|(id, _, _)| id)
        .collect();
    let moved = ids.len();
    state
        .db
        .call(move |conn| store::update_pane_runtimes(conn, &ids, &runtime))
        .await?;
    Ok(moved)
}

/// How many processes a WSL pane's shell accounts for there on its own: the
/// login shell. The session processes `wsl.exe` starts it under belong to
/// root, so the scan cannot read them and does not count them (measured).
const WSL_SHELL_ALONE: usize = 1;

/// Switches running panes that are plain shells - nothing running in them -
/// to `runtime` now: their window sees the new runtime and starts the shell
/// again there (`syncRuntime` in the frontend). A pane running anything, an
/// agent above all, keeps its shell. Returns how many switched.
async fn switch_plain_panes(state: &AppState, runtime: &str) -> Result<usize, WorkspaceError> {
    let panes = state.db.call(|conn| store::terminal_panes(conn)).await?;
    let to_wsl = matches!(Runtime::parse(runtime), Ok(Runtime::Wsl(_)));
    let running: Vec<(String, u32, String)> = panes
        .into_iter()
        .filter(|(_, _, current)| current != runtime)
        .filter(|(_, cwd, _)| !(to_wsl && windows_only(git_file(cwd).as_deref())))
        .filter_map(|(id, _, current)| state.pty.shell_pid(&id).map(|pid| (id, pid, current)))
        .collect();
    if running.is_empty() {
        return Ok(0);
    }
    let plain = tokio::task::spawn_blocking(move || plain_shells(&running))
        .await
        .unwrap_or_default();
    let switched = plain.len();
    let target = runtime.to_owned();
    state
        .db
        .call(move |conn| store::update_pane_runtimes(conn, &plain, &target))
        .await?;
    Ok(switched)
}

/// Which of `running` - pane, shell pid, runtime - are plain shells: Windows'
/// process table read once, each distro asked once. Unknown counts as busy.
fn plain_shells(running: &[(String, u32, String)]) -> Vec<String> {
    let on_windows = running
        .iter()
        .any(|(_, _, runtime)| !matches!(Runtime::parse(runtime), Ok(Runtime::Wsl(_))));
    let procs = on_windows.then(proctree::snapshot).and_then(Result::ok);
    let mut distros: HashMap<String, Option<HashMap<String, usize>>> = HashMap::new();
    running
        .iter()
        .filter(|(pane, shell, runtime)| match Runtime::parse(runtime) {
            Ok(Runtime::Wsl(distro)) => distros
                .entry(distro.clone())
                .or_insert_with(|| wsl::pane_processes(&distro).ok())
                .as_ref()
                .and_then(|counts| counts.get(pane))
                .is_some_and(|count| *count <= WSL_SHELL_ALONE),
            _ => {
                procs
                    .as_ref()
                    .and_then(|procs| proctree::bare(procs, *shell))
                    == Some(true)
            }
        })
        .map(|(pane, _, _)| pane.clone())
        .collect()
}

/// The terminal new panes open in: what was chosen, else Windows.
pub async fn terminal(state: &AppState) -> Result<String, WorkspaceError> {
    let chosen = state.db.call(|conn| store::find_terminal(conn)).await?;
    Ok(chosen.unwrap_or_else(|| Runtime::Windows.to_string()))
}

/// `pane.terminal`: the terminal, and the choices; with `terminal`, chooses it.
/// Every pane moves to it - one not started yet when it starts, a plain shell
/// at once - except one with something running in it, which keeps its shell.
pub async fn choose_terminal(
    state: &AppState,
    args: TerminalArgs,
) -> Result<TerminalView, WorkspaceError> {
    let distros = tokio::task::spawn_blocking(wsl::distros)
        .await
        .unwrap_or_default();
    if let Some(requested) = args.terminal {
        let checked = check(&requested, &distros)?;
        let stored = checked.clone();
        state
            .db
            .call(move |conn| store::update_terminal(conn, &stored))
            .await?;
        let moved = move_panes(state, checked.clone(), true).await?;
        if moved + switch_plain_panes(state, &checked).await? > 0 {
            state.bus.publish("workspaces");
        }
    }
    Ok(TerminalView {
        terminal: terminal(state).await?,
        runtimes: std::iter::once(Runtime::Windows)
            .chain(distros.into_iter().map(Runtime::Wsl))
            .map(|runtime| runtime.to_string())
            .collect(),
    })
}

/// As Dex starts, before any shell has: every terminal pane opens in the
/// chosen terminal, which is what choosing it means. A chosen distro that has
/// since been uninstalled would leave every pane dead, so Dex goes back to
/// Windows, and says so.
pub async fn open_panes_in_terminal(state: &AppState) -> Result<(), WorkspaceError> {
    let mut runtime = terminal(state).await?;
    if let Ok(Runtime::Wsl(distro)) = Runtime::parse(&runtime) {
        let installed = tokio::task::spawn_blocking(wsl::distros)
            .await
            .unwrap_or_default();
        if !installed
            .iter()
            .any(|name| name.eq_ignore_ascii_case(&distro))
        {
            tracing::warn!(%distro, "the chosen WSL distro is not installed; Dex opens in Windows");
            runtime = Runtime::Windows.to_string();
            let windows = runtime.clone();
            state
                .db
                .call(move |conn| store::update_terminal(conn, &windows))
                .await?;
        }
    }
    move_panes(state, runtime, false).await.map(|_| ())
}
